### PR TITLE
scheduler: use must_call for async write callback, add assertion for expected path

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What is changed and how it works?

Issue Number: Ref #14838 

What's Changed:
1. Use `must_call` for the `async_write` callback.
2. Add assertion for the path the task is finished unexpectedly and the latch is not released.

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
  - Inject the failpoint `applied_cb_return_undetermined_err` and raise a write task, tikv-server panics.

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/server/raftkv/mod.rs [hunk 1]
```diff
@@ -205,6 +206,25 @@ pub fn drop_snapshot_callback<T>() -> kv::Result<T> {
     Err(kv::Error::from(kv::ErrorInner::Request(err)))
 }
 
+pub fn async_write_callback_dropped_err() -> errorpb::Error {
```
Comment (overvenus) [inline src/server/raftkv/mod.rs:209]: Could you use a concrete type? The message might be changed anytime.

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -1069,10 +1074,14 @@ impl<E: Engine, L: LockManager> TxnScheduler<E, L> {
...
 
-    fn is_undetermined_error(_e: &tikv_kv::Error) -> bool {
-        // TODO: If there's some cases that `engine.async_write` returns error but it's
-        // still possible that the data is successfully written, return true.
-        false
+    // Return true if raftstore returns error and the underlying write status could
+    // not be decided.
+    fn is_undetermined_error(e: &tikv_kv::Error) -> bool {
+        if let tikv_kv::ErrorInner::Request(error_header) = &*(e.0) {
+            error_header.message == "async write on_applied callback is dropped"
```
Comment (overvenus) [inline src/storage/txn/scheduler.rs:1081]: Could you use a concrete type? The message might be changed anytime.

**Review by ekexium — COMMENTED** [review]
It'd be better to have a unit test.

File: src/storage/txn/scheduler.rs (hunk 2 above)
Comment (cfzjywxk) [reply src/storage/txn/scheduler.rs]: An error type is abstracted, to make it simple the `kvproto` is changed, PTAL.

**Comment by cfzjywxk** [comment]
> It'd be better to have a unit test.

@ekexium 
A test case is added, PTAL.

**Review by ekexium — APPROVED** [review]

**Review by zyguan — APPROVED** [review]

File: src/storage/txn/scheduler.rs [hunk 3]
```diff
@@ -1069,10 +1075,18 @@ impl<E: Engine, L: LockManager> TxnScheduler<E, L> {
             .unwrap();
     }
 
-    fn is_undetermined_error(_e: &tikv_kv::Error) -> bool {
-        // TODO: If there's some cases that `engine.async_write` returns error but it's
-        // still possible that the data is successfully written, return true.
-        false
+    // Return true if raftstore returns error and the underlying write status could
+    // not be decided.
+    fn is_undetermined_error(e: &tikv_kv::Error) -> bool {
```
Comment (MyonKeminta) [inline src/storage/txn/scheduler.rs:1080]: This function is reserved for controlling fair locking's behavior. As the undetermined error is just introduced, I think we may need to consider add more related tests for fair locking later.

**Review by MyonKeminta — APPROVED** [review]
LGTM

File: src/storage/txn/scheduler.rs (hunk 3 above)
Comment (cfzjywxk) [reply src/storage/txn/scheduler.rs:1080]: Good idea.

**Pull request merged**